### ISSUE TITLE
FAQ

### ISSUE THREAD

**Issue by aky15** [thu-spmi/CAT#19]
I have written the answers to some [frequently asked questions](https://github.com/thu-spmi/CAT/blob/master/faq_ch.md) about CAT. It would be nice if others could share some suggestions & experience about CAT. And we welcome those who are interested in CAT to translate FAQs into English, or supplement relevant English documents.

**Comment by Justbreak** [comment]
I made an error during the second step, showing that the other two files were successfully compiled, but the file path_weight was not compiled successfully; what impact will this have on subsequent model training?

**Comment by aky15** [comment]
path_weight is used to calculate the log probability of the label sequence log p(l). In training, log p(l) behaves like a constant, so log p(l) is pre-calculated based on the denominator n-gram LM and saved. If path_weight was not compiled successfully, the loss computed may not reflect the real loss, though it will not affect the forward and back-propagation.

**Comment by Justbreak** [comment]
I am compiling the directory path_weight is an error, has anyone encountered it?   Is it the gcc version?

-------------------------------------------------------------------------------------------------------------
-- Configuring done
-- Generating done
-- Build files have been written to: /home/storage06/lidongbo/workspace/CAT/src/ctc_crf/path_weight/build
cd path_weight/build && make
make[1]: Entering directory `/home/storage06/lidongbo/workspace/CAT/src/ctc_crf/path_weight/build'
make[2]: Entering directory `/home/storage06/lidongbo/workspace/CAT/src/ctc_crf/path_weight/build'
make[3]: Entering directory `/home/storage06/lidongbo/workspace/CAT/src/ctc_crf/path_weight/build'
Scanning dependencies of target path_weight
make[3]: Leaving directory `/home/storage06/lidongbo/workspace/CAT/src/ctc_crf/path_weight/build'
make[3]: Entering directory `/home/storage06/lidongbo/workspace/CAT/src/ctc_crf/path_weight/build'
[100%] Building CXX object CMakeFiles/path_weight.dir/path_weight.cc.o
Linking CXX executable path_weight
CMakeFiles/path_weight.dir/path_weight.cc.o: In function `fst::internal::FstImpl<fst::ArcTpl<fst::TropicalWeightTpl<float> > >::ReadHeader(std::istream&, fst::FstReadOptions const&, int, fst::FstHeader*)':
path_weight.cc:(.text._ZN3fst8internal7FstImplINS_6ArcTplINS_17TropicalWeightTplIfEEEEE10ReadHeaderERSiRKNS_14FstReadOptionsEiPNS_9FstHeaderE[_ZN3fst8internal7FstImplINS_6ArcTplINS_17TropicalWeightTplIfEEEEE10ReadHeaderERSiRKNS_14FstReadOptionsEiPNS_9FstHeaderE]+0x2b6): undefined reference to `fst::FstHeader::Read(std::istream&, std::string const&, bool)'
CMakeFiles/path_weight.dir/path_weight.cc.o: In function `fst::ImplToExpandedFst<fst::internal::VectorFstImpl<fst::VectorState<fst::ArcTpl<fst::TropicalWeightTpl<float> >, std::allocator<fst::ArcTpl<fst::TropicalWeightTpl<float> > > > >, fst::MutableFst<fst::ArcTpl<fst::TropicalWeightTpl<float> > > >::Read(std::string const&)':
path_weight.cc:(.text._ZN3fst17ImplToExpandedFstINS_8internal13VectorFstImplINS_11VectorStateINS_6ArcTplINS_17TropicalWeightTplIfEEEESaIS7_EEEEENS_10MutableFstIS7_EEE4ReadERKSs[_ZN3fst17ImplToExpandedFstINS_8internal13VectorFstImplINS_11VectorStateINS_6ArcTplINS_17TropicalWeightTplIfEEEESaIS7_EEEEENS_10MutableFstIS7_EEE4ReadERKSs]+0x229): undefined reference to `fst::FstReadOptions::FstReadOptions(std::string const&, fst::FstHeader const*, fst::SymbolTable const*, fst::SymbolTable const*)'
path_weight.cc:(.text._ZN3fst17ImplToExpandedFstINS_8internal13VectorFstImplINS_11VectorStateINS_6ArcTplINS_17TropicalWeightTplIfEEEESaIS7_EEEEENS_10MutableFstIS7_EEE4ReadERKSs[_ZN3fst17ImplToExpandedFstINS_8internal13VectorFstImplINS_11VectorStateINS_6ArcTplINS_17TropicalWeightTplIfEEEESaIS7_EEEEENS_10MutableFstIS7_EEE4ReadERKSs]+0x288): undefined reference to `fst::FstReadOptions::FstReadOptions(std::string const&, fst::FstHeader const*, fst::SymbolTable const*, fst::SymbolTable const*)'
CMakeFiles/path_weight.dir/path_weight.cc.o: In function `fst::internal::FstImpl<fst::ArcTpl<fst::TropicalWeightTpl<float> > >::WriteFstHeader(fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float> > > const&, std::ostream&, fst::FstWriteOptions const&, int, std::string const&, unsigned long, fst::FstHeader*)':
path_weight.cc:(.text._ZN3fst8internal7FstImplINS_6ArcTplINS_17TropicalWeightTplIfEEEEE14WriteFstHeaderERKNS_3FstIS5_EERSoRKNS_15FstWriteOptionsEiRKSsmPNS_9FstHeaderE[_ZN3fst8internal7FstImplINS_6ArcTplINS_17TropicalWeightTplIfEEEEE14WriteFstHeaderERKNS_3FstIS5_EERSoRKNS_15FstWriteOptionsEiRKSsmPNS_9FstHeaderE]+0x127): undefined reference to `fst::FstHeader::Write(std::ostream&, std::string const&) const'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::compare(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_create(unsigned long&, unsigned long)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_append(char const*, unsigned long)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_replace(unsigned long, unsigned long, char const*, unsigned long)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `vtable for std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::basic_ofstream<char, std::char_traits<char> >::basic_ofstream(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::_Ios_Openmode)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `vtable for std::__cxx11::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `vtable for std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_mutate(unsigned long, unsigned long, char const*, unsigned long)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::reserve(unsigned long)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `VTT for std::__cxx11::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::resize(unsigned long, char)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)@GLIBCXX_3.4.20'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::_M_sync(char*, unsigned long, unsigned long)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::append(char const*)@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `VTT for std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::find(char const*, unsigned long, unsigned long) const@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::compare(char const*) const@GLIBCXX_3.4.21'
/home/storage06/lidongbo/workspace/kaldi-pytorch/tools/openfst/lib/libfst.so: undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::swap(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)@GLIBCXX_3.4.21'
collect2: error: ld returned 1 exit status
make[3]: *** [path_weight] Error 1

**Comment by aky15** [comment]
Yes. Maybe the gcc you used to compile openfst is not what you are currently using?  Switch gcc or recompile openfst may help.

**Comment by Justbreak** [comment]
I directly put path_weight.cc into Kaldi to compile successfully, it does not seem to be a problem with the gcc version
